app/helpers/foreman_host_reports/reports_dashboard_helper.rb

```diff
@@ -0,0 +1,92 @@
+module ForemanHostReports
+  module ReportsDashboardHelper
+
+    def searchable_links(name, search, counter)
+      # TODO: implement global filtering
+      content_tag :li do
+        content_tag(:span, raw('&nbsp;'), :class => 'label', :style => "background-color:" + report_color[counter]) +
+          raw('&nbsp;') +
+          link_to(name, hosts_path(:search => search), :class => "dashboard-links") +
+          content_tag(:h4, totals[counter])
+      end
+    end
+
+    def report_color
+      {
+        :change_hosts => "#4572A7",
+        :failure_hosts => "#AA4643",
+        :nochange_hosts => "#DB843D",
+        :disabled_hosts => "#92A8CD",
+      }
+    end
+
+    def get_overview(report, options = {})
+      state_labels = {
+        change_hosts: _('Change'),
+        failure_hosts: _('Failure'),
+        nochange_hosts: _('No Change'),
+        disabled_hosts: _('Disabled alerts'),
+      }
+      state_filters = {
+        change_hosts: report_changes.count,
+        nochange_hosts: report_nochanges.count,
+        failure_hosts: report_failures.count,
+        disabled_hosts: disabled_alerts.count
+      }
+
+      {
+        data: state_labels.map { |key, label| [label, report[key], report_color[key]] },
+        searchUrl: hosts_path(search: '~VAL~'),
+        searchFilters: state_filters.each_with_object({}) do |(key, filter), filters|
+          filters[state_labels[key]] = filter
+        end,
+      }
+    end
+
+    def totals
+      @report = {} if @report.nil?
+      @report.update(
+        :total_hosts => Host.all.count,
+        :change_hosts => report_changes.count,
+        :nochange_hosts => report_nochanges.count,
+        :failure_hosts => report_failures.count,
+        :disabled_hosts => disabled_alerts.count
+      )
+    end
```

Comment:
Memorization of a hash can be done in a nicer way I think:

    @report_totals ||= { ... }

Reply:
@ezr-ondrej or @ares can you give me advice here, I was researching this a bit but I am unsure.

My concern is thread safety of member variables in view/controller helpers. This probably stems from my lack of understanding how module including/extending/prepending works in Ruby. Do multiple RoR requests share the helper module therefore this memorization would not be safe? Or does RoR creates new instance of a helper per thread so it's actually safe.

See, widgets don't have plugin API that would allow defining data sources, therefore, we must load them either in the view itself or (probably better) in a helper. Thanks.